pkg/ebpf/bytecode/runtime/runtime_compilation_helpers.go

```diff
@@ -0,0 +1,172 @@
+)
+
+var (
+	defaultFlags = []string{
+		"-DCONFIG_64BIT",
+		"-D__BPF_TRACING__",
+		`-DKBUILD_MODNAME="ddsysprobe"`,
+		"-Wno-unused-value",
+		"-Wno-pointer-sign",
+		"-Wno-compare-distinct-pointer-types",
+		"-Wunused",
+		"-Wall",
+		"-Werror",
+	}
+)
+
+func ComputeFlagsAndHash(additionalFlags []string) ([]string, string) {
+	flags := make([]string, len(defaultFlags)+len(additionalFlags))
+	copy(flags, defaultFlags)
+	copy(flags[len(defaultFlags):], additionalFlags)
+
+	flagHash := hashFlags(flags)
+	return flags, flagHash
+}
+
+func hashFlags(flags []string) string {
+	h := sha256.New()
+	for _, f := range flags {
+		h.Write([]byte(f))
+	}
+	return fmt.Sprintf("%x", h.Sum(nil))
+}
+
+type RuntimeCompilationTelemetry struct {
+	compilationEnabled  bool
+	compilationResult   CompilationResult
+	compilationDuration time.Duration
+	headerFetchResult   kernel.HeaderFetchResult
+}
+
+func NewRuntimeCompilationTelemetry() RuntimeCompilationTelemetry {
+	return RuntimeCompilationTelemetry{
+		compilationEnabled: false,
+		compilationResult:  notAttempted,
+		headerFetchResult:  kernel.NotAttempted,
+	}
+}
+
+func (tm *RuntimeCompilationTelemetry) GetTelemetry() map[string]int64 {
+	stats := make(map[string]int64)
+	if tm.compilationEnabled {
+		stats["runtime_compilation_enabled"] = 1
+		stats["runtime_compilation_result"] = int64(tm.compilationResult)
+		stats["kernel_header_fetch_result"] = int64(tm.headerFetchResult)
+		stats["runtime_compilation_duration"] = tm.compilationDuration.Nanoseconds()
+	} else {
+		stats["runtime_compilation_enabled"] = 0
+	}
+	return stats
+}
+
+func (tm *RuntimeCompilationTelemetry) SendMetrics(client *statsd.Client) error {
+	tags := []string{fmt.Sprintf("version:%s", version.AgentVersion)}
+
+	var enabled float64 = 0
+	if tm.compilationEnabled {
+		enabled = 1
+	}
+	if err := client.Gauge(metrics.MetricRuntimeCompiledConstantsEnabled, enabled, tags, 1); err != nil {
+		return err
+	}
+
+	// if the runtime compilation is not enabled we return directly
+	if !tm.compilationEnabled {
+		return nil
+	}
+
+	if err := client.Gauge(metrics.MetricRuntimeCompiledConstantsCompilationResult, float64(tm.compilationResult), tags, 1); err != nil {
+		return err
+	}
+	if err := client.Gauge(metrics.MetricRuntimeCompiledConstantsCompilationDuration, float64(tm.compilationDuration), tags, 1); err != nil {
+		return err
+	}
+	return client.Gauge(metrics.MetricRuntimeCompiledConstantsHeaderFetchResult, float64(tm.headerFetchResult), tags, 1)
+}
+
+type RuntimeCompilationFileProvider interface {
+	GetInputFilename() string
+	GetInputReader(config *ebpf.Config, tm *RuntimeCompilationTelemetry) (io.Reader, error)
+	GetOutputFilePath(config *ebpf.Config, kernelVersion kernel.Version, flagHash string, tm *RuntimeCompilationTelemetry) (string, error)
+}
+
+func RuntimeCompileObjectFile(config *ebpf.Config, cflags []string, provider RuntimeCompilationFileProvider, tm *RuntimeCompilationTelemetry) (CompiledOutput, error) {
```

Comment:
I like the idea of abstracting the compilation logic away from the `RuntimeAsset` class, however I think this way of doing it is a little confusing for 2 main reasons:
- the `telemetry` object is owned by `RuntimeAsset`, but never directly modified by it - instead, `telemetry` is treated in a functional programming way and passed around from function to function
- the relationship between `RuntimeAsset` and this function is a little convoluted:`RuntimeAsset` uses this helper, but the helper needs access to functions defined by `RuntimeAsset`, but it accesses them through the `RuntimeCompilationFileProvider` interface...

WDYK about having something like this?
```
type RuntimeAsset struct {
	filename string
	hash     string

	compiler runtimeAssetCompiler
}

type fileVerificationFn func(string) (io.Reader, string, error)

type runtimeAssetCompiler struct {
	telemetry RuntimeCompilationTelemetry
}

func (c *runtimeAssetCompiler) compile(inputFilename, outputFilename string, verify fileVerificationFn) {
	...
}
```

Reply:
Thanks for taking a look. I agree with you that the code is a bit convoluted right now. I think reversing the dependency like you are proposing is a good idea, I will try to implement this (or something looking like it at least)